lib/backend.js

```diff
@@ -568,6 +573,80 @@ Backend.prototype.getChannels = function(collection, id) {
   ];
 };
 
+Backend.prototype.getSnapshot = function(agent, index, id, version, timestamp, callback) {
+  var backend = this;
+  this._getSnapshot(agent, index, id, version, timestamp, function (error, snapshot) {
+    if (error) return callback(error);
+
+    var request = {
+      collection: index,
+      id: id,
+      version: snapshot.version,
+      timestamp: snapshot.timestamp,
+      snapshots: snapshot.data ? [snapshot.data] : [],
+      type: snapshot.type
+    };
+
+    backend.trigger(backend.MIDDLEWARE_ACTIONS.readSnapshots, agent, request, function (error) {
+      if (error) return callback(error);
+      callback(null, {
+        data: request.snapshots[0],
+        version: request.version,
+        timestamp: request.timestamp,
+        type: request.type
+      });
+    });
+  });
+};
+
+Backend.prototype._getSnapshot = function (agent, index, id, version, timestamp, callback) {
+  version = (typeof version === 'number' && isFinite(version)) ? version + 1 : null;
+  var options = { metadata: true };
+  var timestampIsNumber = typeof timestamp === 'number';
+
+  this._getOps(agent, index, id, 0, version, options, function (error, ops) {
+    if (error) return callback(error);
+
+    var type;
+    var snapshot;
+    var fetchedTimestamp;
+    var fetchedVersion;
```

Comment:
I'm not sure we should initialise `fetchedTimestamp` and `fetchedVersion` to these values. If the version can't be found, then we'll erroneously tell the consumer that we fetched things with these values, when we didn't.

Reply:
(Also the linter calls me out for initialising to `undefined`, which I think is a fair thing for it to complain about)

Reply:
- `var type = null;` - I suggested it only to simplify the implicit type of the variable `null | |OtType`, rather than `undefined | null | OtType`. Not a big deal either way.
- `var snapshot = undefined;` - just to be explicit. I don't mind - it can be left unassigned.
- `var fetchedVersion = 0;` - I think the version should be 0 for nonexistent snapshots for consistency with `Doc` and https://github.com/share/sharedb#data-model, especially this part of the docs: "ShareDB implicitly has a record for every document you can access.". I think it's also a good practice to stick to a single data type for each variable whenever possible. In the case of `version` I think it is possible and quite natural to have `version === 0` for nonexistent snapshots - the same as `Doc.version` for nonexistent docs.
- `var fetchedTimestamp = timestamp;` - similar to `fetchedVersion` - keep the data type the same (number). The user would now that the snapshot does not exist by checking that `version === 0`. Actually, it might be better to do `var fetchedTimestamp = 0;`, in case `timestamp` is not specified. Additionally, `timestamp` would be meaningless for nonexistent snapshots anyway.

Reply:
> I think it is possible and quite natural to have version === 0 for nonexistent snapshots

Reply:
For what it's worth, the other reason I don't like this is that there's no easy way to tell if the snapshot you got back is valid. For example, if you asked for a `timestamp`, you could check if `timestamp === 0`, but (I suppose) you could have asked for `timestamp: 0`. You could also check `type === null`, but this is `true` if the snapshot is deleted, but still valid. But if this is what you want, I'll get a change pushed.

Reply:
>  It's possible to call this method and get two snapshots claiming to be version === 0, but with different data.

I don't see how it would be possible - `version` is `0` only if we don't process any ops, and in that case `data` is always `undefined`.

> For what it's worth, the other reason I don't like this is that there's no easy way to tell if the snapshot you got back is valid. For example, if you asked for a timestamp, you could check if timestamp === 0, but (I suppose) you could have asked for timestamp: 0. You could also check type === null, but this is true if the snapshot is deleted, but still valid. But if this is what you want, I'll get a change pushed.

*I assume that by "valid snapshot" you mean that a document existed at a specific point in time.*

Your reasoning above is correct and the only reliable way to know that a document did not exist at a particular time is to check for `version === 0`. So, you still get the same information as before the change I suggested but in a slightly different way.

Reply:
> version is 0 only if we don't process any ops, and in that case data is always undefined

The creation op is v0.

Reply:
Ok, I think I get it now... op version `X` in the database means that the op can be applied to a doc at version `X`. After the op is applied, the doc is at version `X + 1`. I think `Backend.prototype._getSnapshot` will need to be updated to reflect that, then we should both be happy. :-)

Reply:
Riiiight. So creation is v1?

Reply:
The initial `create` is applied to a doc at version 0, so after it's applied, the doc is v1.

Reply:
Great. Pushed a change that matches this logic. I'm now much happier about v0 being a non-existent doc.

Reply:
Nice, thanks @alecgibson :+1: